Set proper default command output

RiccardoM opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description
This PR sets the proper default command output

closes: #8498

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

amaurym commented on simapp/simd/cmd/root.go:
```diff
@@ -83,6 +83,12 @@ func initRootCmd(rootCmd *cobra.Command, encodingConfig params.EncodingConfig) {
 
+	// set the default command output properly
+	rootCmd.PersistentPreRun = func(cmd *cobra.Command, args []string) {
+		cmd.SetOut(cmd.OutOrStdout())
```
Doing a search of "cmd.SetOut(cmd.OutOrStdout())", we do it in several places. Do you think we can remove all those instances, and just keep this?

RiccardoM replied:
@amaurym Thanks for the heads up. I've removed all the duplicated calls.

amaurym reviewed (commented):
lgtm, though I didn't test it. @RiccardoM did you test it locally on your machine?

Also, a changelog entry would be nice. I believe this is considered client-breaking.

RiccardoM commented:
> lgtm, though I didn't test it. @RiccardoM did you test it locally on your machine?
> 
> Also, a changelog entry would be nice. I believe this is considered client-breaking.

Added the CHANGELOG. I also moved the place where the outputs were initialized cause it didn't work. The new method using the already existing `PersistentPreRunE` instead of `PersistentPreRun` now works properly.

alessio reviewed (changes requested):
Hold off please. Errors and diagnostics always go to stderr. When a command is meant to produce an output, that should go to stdout. Don't change defaults please.

RiccardoM commented:
> Hold off please. Errors and diagnostics always go to stderr. When a command is meant to produce an output, that should go to stdout. Don't change defaults please.

Shouldn't this be a responsibility of commands? IMO they should call `cmd.PrintErrln()` for errors, not `cmd.Println`. It would make more sense from my point of view to use the proper methods instead of having the default output to `stderr`

alessio commented:
> > Hold off please. Errors and diagnostics always go to stderr. When a command is meant to produce an output, that should go to stdout. Don't change defaults please.
> 
> IMO they should call `cmd.PrintErrln()` for errors, not `cmd.Println`.

Yes, I agree. It makes sense. I agree with @alexanderbez, we need some grooming here. Just changing defaults seems the most obvious way to get the wrong solution in place.

alessio commented on client/flags/flags.go:
```diff
@@ -88,9 +88,6 @@ func AddQueryFlagsToCmd(cmd *cobra.Command) {
 	cmd.Flags().StringP(tmcli.OutputFlag, "o", "text", "Output format (text|json)")
 
 	cmd.MarkFlagRequired(FlagChainID)
-
-	cmd.SetErr(cmd.ErrOrStderr())
-	cmd.SetOut(cmd.OutOrStdout())
```
Can we revert this, please? This might change the defaults

alessio commented on client/flags/flags.go:
```diff
@@ -119,9 +116,6 @@ func AddTxFlagsToCmd(cmd *cobra.Command) {
 	cmd.Flags().String(FlagGas, "", fmt.Sprintf("gas limit to set per-transaction; set to %q to calculate sufficient gas automatically (default %d)", GasFlagAuto, DefaultGasLimit))
 
 	cmd.MarkFlagRequired(FlagChainID)
-
-	cmd.SetErr(cmd.ErrOrStderr())
-	cmd.SetOut(cmd.OutOrStdout())
```
As above: can we revert this, please? As it might change the defaults

alessio commented on client/keys/add.go:
```diff
@@ -77,9 +77,6 @@ the flag --nosort is set.
 	cmd.Flags().Uint32(flagIndex, 0, "Address index number for HD derivation")
 	cmd.Flags().String(flags.FlagKeyAlgorithm, string(hd.Secp256k1Type), "Key signing algorithm to generate keys for")
 
-	cmd.SetOut(cmd.OutOrStdout())
-	cmd.SetErr(cmd.ErrOrStderr())
-
```
Ditto

alessio replied:
OK. Just make sure that default behaviour is not affected :+1: i.e. test manually, please

michaelfig replied:
> OK. Just make sure that default behaviour is not affected 👍 i.e. test manually, please

I went through and manually tested, as well as audited all the uses of `cmd.Print*`.  That's what https://github.com/cosmos/cosmos-sdk/pull/8628/commits/8c78b32ba21bc63e6056b909cd23559feb0b8fbf is the result of.

alessio commented on client/keys/list.go:
```diff
@@ -34,8 +34,6 @@ func runListCmd(cmd *cobra.Command, _ []string) error {
 		return err
 	}
 
-	cmd.SetOut(cmd.OutOrStdout())
-
```
Ditto

alessio commented on server/export.go:
```diff
@@ -109,8 +109,7 @@ func ExportCmd(appExporter types.AppExporter, defaultNodeHome string) *cobra.Com
 	}
-	cmd.SetOut(cmd.OutOrStdout())
-	cmd.SetErr(cmd.ErrOrStderr())
+
```
Ditto

michaelfig commented:
It is unfortunate that this PR has not gained any traction.  It is a major ergonomic failure (and violation of standard *nix pipelineable commands) to have machine-readable output go to stderr.  If I run `appd status | jq ...`, the output is not captured.  If I run `appd status 2>&1 | jq ...` then error messages are combined with the output and are hidden from view if there is a failure.

@alessio, what will it take to merge?

alessio commented:
> It is unfortunate that this PR has not gained any traction. It is a major ergonomic failure (and violation of standard *nix pipelineable commands) to have machine-readable output go to stderr. If I run `appd status | jq ...`, the output is not captured. If I run `appd status 2>&1 | jq ...` then error messages are combined with the output and are hidden from view if there is a failure.
> 
> @alessio, what will it take to merge?

I'm testing this again now.

alessio commented:
@RiccardoM mind resolving the conflicts, please?

RiccardoM commented:
> @RiccardoM mind resolving the conflicts, please?

Should be done

amaurym commented:
tACK. Tested with a couple of CLI subcommands outputting JSON, and piping to jq

amaurym commented:
I believe backporting this could be useful for gaia CLI users.

shahankhatch reviewed (approved):
Thanks! Very helpful for gaia users.

michaelfig commented:
Thank you, everybody!
